wallet: ensure the wallet is unlocked when needed for rescanning

ishaanam opened the pull request:
Wallet passphrases are needed to top up the keypool of encrypted wallets
during a rescan. The following RPCs need the passphrase when rescanning:
    - `importdescriptors`
    - `rescanblockchain`

The following RPCs use the information about whether or not the
passphrase is being used to ensure that full rescans are able to
take place (meaning the following RPCs should not be able to run
if a rescan requiring the wallet to be unlocked  is taking place):
    - `walletlock`
    - `encryptwallet`
    - `walletpassphrasechange`
  
`m_relock_mutex` is also introduced so that the passphrase is not
deleted from memory when the timeout provided in
`walletpassphrase` is up and the wallet is still rescanning.
Fixes #25702, #11249

Thanks to achow101 for coming up with the idea of using a new mutex to solve this issue and for answering related questions.

ishaanam committed:
wallet: keep track of when the passphrase is needed when rescanning

Wallet passphrases are needed to top up the keypool during a
rescan. The following RPCs need the passphrase when rescanning:
    - `importdescriptors`
    - `rescanblockchain`

The following RPCs use the information about whether or not the
passphrase is being used to ensure that full rescans are able to
take place:
    - `walletlock`
    - `encryptwallet`
    - `walletpassphrasechange`

aureleoules reviewed (commented):
Changes look good, is there a reason it's in draft?
Also, is there a way to add a test for this?

ishaanam commented:
> Changes look good, is there a reason it's in draft?
> 
> Also, is there a way to add a test for this?

I made it a draft because I haven't written a test yet. I'll mark it as ready for review after I add some tests.

aureleoules commented:
The test seems to create intermittent failures. I also cherry-picked the test on master and the results seem to be the same.
`seq 100 | xargs -i -P 5 python test/functional/wallet_importdescriptors.py --descriptors` will execute the test in parallel batches of 5.

ishaanam commented:
> The test seems to create intermittent failures. I also cherry-picked the test on master and the results seem to be the same. `seq 100 | xargs -i -P 5 python test/functional/wallet_importdescriptors.py --descriptors` will execute the test in parallel batches of 5.

Thanks for taking a look at this, that should be fixed now. There is still a problem of the passphrase timing out on certain systems before `importdescriptors` is even run, but I'm not sure how to fix that.

achow101 commented on src/wallet/rpc/encrypt.cpp:
```diff
@@ -90,6 +90,7 @@ RPCHelpMan walletpassphrase()
     std::weak_ptr<CWallet> weak_wallet = wallet;
     pwallet->chain().rpcRunLater(strprintf("lockwallet(%s)", pwallet->GetName()), [weak_wallet, relock_time] {
         if (auto shared_wallet = weak_wallet.lock()) {
+            LOCK(shared_wallet->m_relock_mutex);
```
In b8952bd7d3b3f745109506223f829c5dd91e638f "wallet: ensure that the passphrase is not deleted from memory when being used to rescan"

I think this is the wrong place to grab this mutex. I think it would be better to put it `CWallet::Lock` (in addition to the `LOCK(cs_wallet)` already there) as that is the underlying locking function that is called in a variety of places and so would offer stronger protection.

hernanmarino replied:
I have some doubts about the proper location for grabbing the mutex, but I am not an expert in wallet code, some take my ideas with care. @achow101 's suggestion seems like a nice "natural" way to lock mutexes, and initially I agreed with it. But since this pr creates a new mutex, I see no conflict on grabbing it elsewhere, like here. Also I'm a little bit concerned that other calls to `CWallet::Lock` might unnecessarily lock this unrelated mutex, blocking execution.

ishaanam replied:
The new mutex is now grabbed in `CWallet::Lock` and also in locations where `cs_wallet` would have been grabbed first in order to prevent a deadlock.

achow101 commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -875,6 +875,7 @@ RPCHelpMan rescanblockchain()
     uint256 start_block;
 
     EnsureWalletIsUnlocked(*pwallet);
+    LOCK(pwallet->m_relock_mutex);
```
In b8952bd7d3b3f745109506223f829c5dd91e638f "wallet: ensure that the passphrase is not deleted from memory when being used to rescan"

The mutex should be grabbed before ensuring the wallet is unlocked. There could be a race condition where the wallet is locked by the scheduler in between checking that the wallet is unlocked, and grabbing `m_relock_mutex`. By grabbing the mutex first, we can be sure that the scheduler will not be able to lock the wallet before we check that the wallet is unlocked.

Also, use `LOCK2` to grab 2 locks at the same time.

ishaanam replied:
How would using `LOCK2` work for grabbing `cs_wallet` and `m_relock_mutex` at the same time if the whole point of having `m_relock_mutex` is that we don't want to be holding `cs_wallet` during the time of the rescan since it can be time consuming?

achow101 replied:
`LOCK2` only needs to be used in the places when both locks are being acquired, e.g. in `CWallet::Lock`.

ishaanam replied:
I have made it so that `m_relock_mutex` is grabbed before ensuring the wallet is unlocked.
> LOCK2 only needs to be used in the places when both locks are being acquired, e.g. in CWallet::Lock.

Right, I thought you meant this location specifically.

achow101 commented on test/functional/wallet_importdescriptors.py:
```diff
@@ -666,5 +666,28 @@ def run_test(self):
+        descriptor = {"desc": descsum_create("pkh(" + xpriv + "/1h/*h)"),
+                              "timestamp": "now",
+                              "active": True,
+                              "range": [0,10000],
+                              "next_index": 10000}
```
In ad8d139d6237ffa556e90d2d87ec43c209c89dc9 "test: test rescanning encrypted wallets"

10000 is running into the rpc timeout for me for the `importdescriptors` calls. I think this could be reduced to 4000. That seems to not hit the rpc timeout, but still consistently fails this test on master.

hernanmarino replied:
10000 runs fine for me, but I agree with lowering it to 4000.

ishaanam replied:
Done

achow101 reviewed (commented):
This could have a similar test for `rescanblockchain` as well.

pablomartin4btc reviewed (commented):
Concept ACK, I see the point from @achow101 moving the wallet lock/ mutex to `CWallet::Lock()`, it looks like a better place but I'd need more knowledge to give a proper advice.

ishaanam committed:
test: test rescanning encrypted wallets

ishaanam committed:
wallet: ensure that the passphrase is not deleted from memory when being used to rescan

`m_relock_mutex` is introduced so that the passphrase is not
deleted from memory when the timeout provided in
`walletpassphrase` is up, but the wallet is still rescanning.

ishaanam commented:
Thanks for the review @achow101, I have implemented your suggestions. I also made some slight modifications to the error messages in 8c16fdb73dd323245642d2710b5f81776db40396.
> This could have a similar test for rescanblockchain as well.

Yes, I will add one shortly.

w0xlt commented on src/wallet/wallet.h:
```diff
@@ -487,6 +487,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     // Used to prevent concurrent calls to walletpassphrase RPC.
     Mutex m_unlock_mutex;
+    // Used to prevent deleting the passphrase from memory when it is still in use.
+    RecursiveMutex m_relock_mutex;
```
Using `Mutex` is preferable to `RecursiveMutex`.

ishaanam replied:
`Mutex` isn't used here because sometimes `m_relock_mutex` has to be locked multiple time within a scope to prevent deadlocks. See: https://github.com/bitcoin/bitcoin/pull/26347#discussion_r1006353769

> The new mutex is now grabbed in CWallet::Lock and also in locations where cs_wallet would have been grabbed first in order to prevent a deadlock.

ishaanam commented:
Added a test for `rescanblockchain`

theStack commented:
Concept ACK

furszy commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -866,15 +866,17 @@ RPCHelpMan rescanblockchain()
+
     {
         LOCK(pwallet->cs_wallet);
+        EnsureWalletIsUnlocked(*pwallet);
```
if the wallet is a watch-only wallet, do we need to ensure this?

ishaanam replied:
No, I will make it so that here and in the other RPCs `EnsureWalletIsUnlocked` is only called if the wallet is not watch-only. Additionally, in that case `m_relock_mutex` doesn't need to be held, as long as it is not grabbed later in the RPC.

ishaanam replied:
Actually, the wallet will always be unlocked if private keys are disabled, because wallet encryption is not allowed in a watch-only wallet. See: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpc/encrypt.cpp#L224

pablomartin4btc reviewed (commented):
re-ack b2e2829014c5c9a3dc02eab7fcd77dd907228c78. I'll check again once you make the changes you commented to @furszy.

hernanmarino reviewed (commented):
Re ACK b2e2829014c5c9a3dc02eab7fcd77dd907228c78

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3284,7 +3285,7 @@ bool CWallet::Lock()
 
     {
-        LOCK(cs_wallet);
+        LOCK2(m_relock_mutex, cs_wallet);
```
Was surprised to see that this wasn't causing any lock ordering deadlock. Because, for example, `ChangeWalletPassphrase` first locks `cs_wallet` and then call `Lock()` (which locks `m_relock_mutex` first, which is the opposite lock ordering direction).

So.. got deeper over it and was able to trigger the deadlock detection on the GUI "change wallet passphrase" workflow.
You can easily replicate it by enabling the debug mode (--enable-debug at configure time) and try to change the wallet passphrase there.

This is not happening on the RPC command merely because this two recursive locks are acquired at the `ChangeWalletPassphrase` caller, which locks them in the proper order (which does not happen on the GUI's `changeWalletPassphrase` interface method).

ishaanam replied:
Good catch, I was able to replicate this locally. Should be fixed now.

pablomartin4btc reviewed (commented):
re-ack fcaaa5fd905a21298bd69343c9756e32a70870e6 - fix lock ordering deadlock.

achow101 commented on test/functional/wallet_importdescriptors.py:
```diff
@@ -666,5 +666,28 @@ def run_test(self):
+
+        self.nodes[0].createwallet("encrypted_wallet", blank=True, descriptors=True, passphrase="passphrase")
+        encrypted_wallet = self.nodes[0].get_wallet_rpc("encrypted_wallet")
+        encrypted_wallet.walletpassphrase("passphrase", 1)
+
+        descriptor["timestamp"] = 0
+        descriptor["next_index"] = 0
+        encrypted_wallet.importdescriptors([descriptor])
```
In fcaaa5fd905a21298bd69343c9756e32a70870e6 "test: test rescanning encrypted wallets"

I suspect that this test is going to be the source of many intermittent failures in the future. We already observe a failure in CI for this test. Since this is testing a race condition, the speed of the machine executing the test becomes a significant factor, and having sanitizers being run can cause the test to fail.

I think we might be able to workaround this by using a batched RPC request as IIRC those are run in a single thread sequentially.

ishaanam replied:
I agree that this test would have been problematic. I have changed the `importdescriptors` and `rescanblockchain` tests added in this PR to use a batched RPC request.

ishaanam commented:
Rebased, I also increased the `walletpassphrase` timeout in both of the tests.

achow101 commented:
ACK 6a5b348f2e526f048d0b448b01f6c4ab608569af

hernanmarino reviewed (commented):
ACK 6a5b348f2e526f048d0b448b01f6c4ab608569af

furszy reviewed (approved):
Tested ACK 6a5b348f

No longer crashes due the deadlock.

ishaanam commented on test/functional/wallet_transactiontime_rescan.py:
```diff
@@ -167,6 +174,38 @@ def run_test(self):
+            minernode.createwallet("encrypted_wallet", blank=True, passphrase="passphrase", descriptors=False)
+            encrypted_wallet = minernode.get_wallet_rpc("encrypted_wallet")
+
+            encrypted_wallet.walletpassphrase("passphrase", 1)
```
> Maybe the timeout can be increased to 999999?

Yes, the timeout before `sethdseed` can be increased to 999999, followed by running `walletlock`. I can open a follow-up PR for this.
> Also I wonder why you picked a timeout of 3 below, or why walletpassphrase needs to be called at all twice in a row?

3 was used because it needed to be a number that was large enough so that the scan would begin, but small enough so that the timeout would occur while the rescan is in progress. However, this number can be too large or too small depending on the speed of the machine.
> If you want to confirm that the wallet can't be relocked while the rescan is in progress, wouldn't it be better to call walletlock rpc right after the rescan rpc?

I don't completely understand this suggestion. Do you mean calling `walletlock` to ensure that the wallet is still unlocked? If so, I think there are two problems with that:

- `walletlock` doesn't throw an error if the wallet is already locked
- The wallet is supposed to re-lock immediately after the rescan, so the wallet might already be locked by the time `walletlock` is called, even if it did not lock during the rescan. However, using a batched RPC request might help with this.

ishaanam replied:
Ok, I think I understand what you mean now. I have implemented this using multi-threading in #27199. However, for me this no longer fails without the fix.

Ronnie6464 reviewed (changes requested):
No

Ronnie6464 commented on test/functional/wallet_transactiontime_rescan.py:
```diff
@@ -167,6 +174,38 @@ def run_test(self):
+            encrypted_wallet = minernode.get_wallet_rpc("encrypted_wallet")
+
+            encrypted_wallet.walletpassphrase("passphrase", 1)
+            encrypted_wallet.sethdseed(seed=hd_seed)
```
Ronowens6565@gmail.com

Ronnie6464 replied:
N[]()

Ronnie6464 commented on test/functional/wallet_transactiontime_rescan.py:
```diff
@@ -167,6 +174,38 @@ def run_test(self):
+            encrypted_wallet = minernode.get_wallet_rpc("encrypted_wallet")
+
+            encrypted_wallet.walletpassphrase("passphrase", 1)
+            encrypted_wallet.sethdseed(seed=hd_seed)
```
- **[]()
```suggestion
            encrypted_wallet.sethdseed(seed=hd_seed)_[]()![git_refs_test.go.txt](https://github.com/bitcoin/bitcoin/files/11583540/git_refs_test.go.txt)

_
```
**

Ronnie6464 reviewed (approved):
[]()

Ronnie6464 commented:
> Wallet passphrases are needed to top up the keypool of encrypted wallets
> during a rescan. The following RPCs need the passphrase when rescanning:
>     - `importdescriptors`
>     - `rescanblockchain`
> 
> The following RPCs use the information about whether or not the
> passphrase is being used to ensure that full rescans are able to
> take place (meaning the following RPCs should not be able to run
> if a rescan requiring the wallet to be unlocked  is taking place):
>     - `walletlock`
>     - `encryptwallet`
>     - `walletpassphrasechange`
>   
> `m_relock_mutex` is also introduced so that the passphrase is not
> deleted from memory when the timeout provided in
> `walletpassphrase` is up and the wallet is still rescanning.
> Fixes #25702, #11249
> 
> Thanks to achow101 for coming up with the idea of using a new mutex to solve this issue and for answering related questions.

Ronnie6464 commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1651,10 +1651,14 @@ RPCHelpMan importdescriptors()
 
     WalletRescanReserver reserver(*pwallet);
-    if (!reserver.reserve()) {
+    if (!reserver.reserve(/*with_passphrase=*/true)) {
```
test/functional/wallet_transactiontime_rescan.py

Ronnie6464 reviewed (approved):
test/functional/wallet_transactiontime_rescan.py
